pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto

```diff
@@ -199,6 +202,13 @@ enum NamespaceMode {
     // For example, a container with a PID namespace of NODE expects to view
     // all of the processes on the host running the kubelet.
     NODE      = 2;
+    // A NODE_WIDE_REMAPPED namespace is common to all the pods on the kubernetes node
+    // and this namespace has uids/gids mapped to a different range of uids/gids on the
+    // kubernetes node namespace.
+    // For example, starting from uid/gid 0, 10000 uids/gids in this namespace are mapped to
+    // 10000 ids on node namespace,starting with id 200000. i.e uid/gid 0 in this namespace is
+    // mapped to uid/gid 200000 on node namespace.
+    NODE_WIDE_REMAPPED = 3;
```

Comment:
is this a valid namespace for the existing network/pid/ipc namespaces? do they all support all existing values in the enum?

Reply:
is this the only mode supported ? nit: the wording is confusing. Something like
``` A NODE_WIDE_REMAPPED namespace applies to all pods on a given kubernetes node. The uid/gids of the pods/containers on a given node are mapped to a range of uids/gids from that Node's namespace. For e.g. starting with 200000 id on the nodes namespace, 10000 ids are allocated for a pod. This means uid/gid 0 inside the pod would map to 200000 id on the nodes namespace. 
```

Reply:
I am assuming each pod would get a different range from the host mapped to its namespace. Is that understanding right ?

Reply:
> I am assuming each pod would get a different range from the host mapped to its namespace. Is that understanding right?

No. I was surprised by this as well. uid 0 inside any pod's container will map to the same host uid.

Reply:
We can do this in CRI-O, But Kubernetes has no way of telling CRI-O yet to pick different UserNS for each pod.  This is the long term goal.

Reply:
if multiple pods on a given host get mapped to the same uid on host, what is the security impact of this ? It seems in this mode, we should add validation in kubelet to say all pods must run as different uids otherwise we are violating the guarantee that each pod has its own namespace. Now multiple pods belong to the same id namespace.

Reply:
That is what we have now, exept the ID Namespace is the hosts UIDs.  IE root==realroot.
For the first step of UserNS we can increase the security by turning the UserNS in the CRIs.
For example if we turn on CRI-O support for UserNS we end up with all the PODs continuing to run in the same USERNS, except now container root != realroot,  and container root has no capabilities.  It is only able to act within its namespace.  If container root breaks out it would be able to attack other containers, based on UID Mappings, but it would have a harder time since it is not root on the system.  Of course SELinux would prevent this on a different level.

In the long run we want to support multiple USERNS, but baby steps...

Reply:
@liggitt No, NODE_WIDE_REMAPPED makes sense only for user namespaces. Initially i was using "POD" to represent user-namespace remapping but because of your concern that currently user-namespace remapping is not  actually pod-level, which is true, this new enum value i introduced to bring clarity.

Reply:
thanks for the explanation @rhatdan i did know that currently container uids were same as host uids and hence had the exact same privilege. This would mean if we are mixing pods from different customers, they must run with different uids currently.